README.md
# here_routing

Asynchronous Python client for the HERE Routing V8 API

[![GitHub Actions](https://github.com/eifinger/here_routing/workflows/CI/badge.svg)](https://github.com/eifinger/here_routing/actions?workflow=CI)
[![PyPi](https://img.shields.io/pypi/v/here_routing.svg)](https://pypi.python.org/pypi/here_routing)
[![PyPi](https://img.shields.io/pypi/l/here_routing.svg)](https://github.com/eifinger/here_routing/blob/master/LICENSE)
[![codecov](https://codecov.io/gh/eifinger/here_routing/branch/master/graph/badge.svg)](https://codecov.io/gh/eifinger/here_routing)
[![Downloads](https://pepy.tech/badge/here_routing)](https://pepy.tech/project/here_routing)

## Installation

```bash
pip install here_routing
```

## Usage

```python
import asyncio

from here_routing import HERERoutingApi, Place, Return, TransportMode

API_KEY = "<YOUR_API_KEY>"


async def main() -> None:
    """Show example how to get duration of your route."""
    async with HERERoutingApi(api_key=API_KEY) as here_routing:
        response = await here_routing.route(
            transport_mode=TransportMode.CAR,
            origin=Place(latitude=50.12778680095556, longitude=8.582081794738771),
            destination=Place(latitude=50.060940891421765, longitude=8.336477279663088),
            return_values=[Return.SUMMARY],
        )
        print(
            f"Duration is: {response['routes'][0]['sections'][0]['summary']['duration']}"
        )


if __name__ == "__main__":
    asyncio.run(main())
```


CHANGELOG.md
# Changelog

<!--next-version-placeholder-->

## v1.0.0 (2023-03-08)
### Fix
* Bump versions and require python >3.10 ([#36](https://github.com/eifinger/here_routing/issues/36)) ([`1e1f0f1`](https://github.com/eifinger/here_routing/commit/1e1f0f1993b41bd4643cec25957e699576cc53c2))

### Breaking
* Require Python3.10 or later ([`1e1f0f1`](https://github.com/eifinger/here_routing/commit/1e1f0f1993b41bd4643cec25957e699576cc53c2))

## v0.2.0 (2022-12-12)
### Feature
* Add HERERoutingTooManyRequestsError ([`9e0af11`](https://github.com/eifinger/here_routing/commit/9e0af11009df7586d06d431319ac410d30d76681))

## v0.1.1 (2022-10-24)
### Fix
* Handle errors with HTTP 200 response code ([`8114ad4`](https://github.com/eifinger/here_routing/commit/8114ad4eb1c308f39c589abe5baf9478a3c5cece))

### Documentation
* Remove weenect from docs ([`8a0bae6`](https://github.com/eifinger/here_routing/commit/8a0bae6dad27d3693d36cd5042d1d2c90056346d))

## v0.1.0 (2022-07-15)


## v1.1.0 (2022-07-13)
### Feature
* Initial commit ([`d530178`](https://github.com/eifinger/here_routing/commit/d53017867acad90ddc9ff2f9c2627b39ad181ce1))


tests/__init__.py
"""Tests for Asynchronous Python client for the HERE Routing V8 API."""


tests/test_here_routing.py
"""Tests for `here_routing.here_routing`."""
import os
from datetime import datetime

import aiohttp
import pytest

from here_routing.exceptions import (
    HERERoutingError,
    HERERoutingTooManyRequestsError,
    HERERoutingUnauthorizedError,
)
from here_routing.here_routing import API_HOST, API_VERSION, ROUTES_PATH, HERERoutingApi
from here_routing.model import Place, Return, Spans, TransportMode


@pytest.mark.asyncio
async def test_car_route(aresponses):
    """Test getting user information with a timed out token."""
    aresponses.add(
        API_HOST,
        f"{API_VERSION}/{ROUTES_PATH}",
        "GET",
        aresponses.Response(
            text=load_json_fixture("car_route_response.json"),
            status=200,
            content_type="application/json",
        ),
    )
    async with aiohttp.ClientSession() as session:
        here_api = HERERoutingApi(api_key="key", session=session)
        response = await here_api.route(
            transport_mode=TransportMode.CAR,
            origin=Place(latitude=50.12778680095556, longitude=8.582081794738771),
            destination=Place(latitude=50.060940891421765, longitude=8.336477279663088),
            return_values=[Return.SUMMARY, Return.POLYINE],
            spans=[Spans.NAMES],
            via=[Place(latitude=50.12778630095556, longitude=8.582081734738771)],
            departure_time=datetime.now(),
        )

        assert response["routes"][0]["sections"][0]["summary"]["duration"] == 1085


@pytest.mark.asyncio
async def test_invalid_key(aresponses):
    """Test that an invalid api_key throws HERERoutingUnauthorizedError."""
    aresponses.add(
        API_HOST,
        f"{API_VERSION}/{ROUTES_PATH}",
        "GET",
        aresponses.Response(
            text=load_json_fixture("invalid_key_response.json"),
            status=401,
            content_type="application/json",
        ),
    )
    async with aiohttp.ClientSession() as session:
        with pytest.raises(HERERoutingUnauthorizedError):
            here_api = HERERoutingApi(api_key="invalid", session=session)
            await here_api.route(
                transport_mode=TransportMode.CAR,
                origin=Place(latitude=50.12778680095556, longitude=8.582081794738771),
                destination=Place(
                    latitude=50.060940891421765, longitude=8.336477279663088
                ),
            )


@pytest.mark.asyncio
async def test_malformed_request(aresponses):
    """Test that a malformed request throws HERERoutingError."""
    aresponses.add(
        API_HOST,
        f"{API_VERSION}/{ROUTES_PATH}",
        "GET",
        aresponses.Response(
            text=load_json_fixture("malformed_request_response.json"),
            status=400,
            content_type="application/json",
        ),
    )
    async with aiohttp.ClientSession() as session:
        with pytest.raises(HERERoutingError):
            here_api = HERERoutingApi(api_key="key", session=session)
            await here_api.route(
                transport_mode=TransportMode.CAR,
                origin=Place(latitude=50.12778680095556, longitude=8.582081794738771),
                destination=Place(
                    latitude=50.060940891421765, longitude=8.336477279663088
                ),
            )


@pytest.mark.asyncio
async def test_invalid_request(aresponses):
    """Test that a invalid request throws HERERoutingError."""
    aresponses.add(
        API_HOST,
        f"{API_VERSION}/{ROUTES_PATH}",
        "GET",
        aresponses.Response(
            text=load_json_fixture("invalid_request_response.json"),
            status=400,
            content_type="application/json",
        ),
    )
    async with aiohttp.ClientSession() as session:
        with pytest.raises(HERERoutingError) as error:
            here_api = HERERoutingApi(api_key="key", session=session)
            await here_api.route(
                transport_mode=TransportMode.CAR,
                origin=Place(latitude=50.12778680095556, longitude=8.582081794738771),
                destination=Place(
                    latitude=50.060940891421765, longitude=8.336477279663088
                ),
            )
        assert "Spans requested but no polyline requested" in str(error.value)


@pytest.mark.asyncio
async def test_429_too_many_requests(aresponses):
    """Test that a invalid request throws HERETransitTooManyRequestsError."""
    aresponses.add(
        API_HOST,
        f"{API_VERSION}/{ROUTES_PATH}",
        "GET",
        aresponses.Response(
            text=load_json_fixture("too_many_requests_response.json"),
            status=429,
            content_type="application/json",
        ),
    )
    async with aiohttp.ClientSession() as session:
        with pytest.raises(HERERoutingTooManyRequestsError) as error:
            here_api = HERERoutingApi(api_key="key", session=session)
            await here_api.route(
                transport_mode=TransportMode.CAR,
                origin=Place(latitude=150.12778680095556, longitude=8.582081794738771),
                destination=Place(
                    latitude=50.060940891421765, longitude=8.336477279663088
                ),
            )
        assert "Rate limit for this service has been reached" in str(error.value)


@pytest.mark.asyncio
async def test_no_route_found(aresponses):
    """Test that a no route found response throws HERERoutingError."""
    aresponses.add(
        API_HOST,
        f"{API_VERSION}/{ROUTES_PATH}",
        "GET",
        aresponses.Response(
            text=load_json_fixture("no_route_response.json"),
            status=200,
            content_type="application/json",
        ),
    )
    async with aiohttp.ClientSession() as session:
        with pytest.raises(HERERoutingError) as error:
            here_api = HERERoutingApi(api_key="key", session=session)
            await here_api.route(
                transport_mode=TransportMode.CAR,
                origin=Place(latitude=50.12778680095556, longitude=8.582081794738771),
                destination=Place(
                    latitude=50.060940891421765, longitude=8.336477279663088
                ),
            )
        assert "Couldn't find a route" in str(error.value)


def load_json_fixture(filename: str) -> str:
    """Load a fixture."""
    path = os.path.join(os.path.dirname(__file__), "fixtures", filename)
    with open(path, encoding="utf-8") as fptr:
        return fptr.read()


here_routing/model.py
"""HERE Routing API models."""

from dataclasses import dataclass
from enum import Enum


class TransportMode(Enum):
    """Available TransportModes."""

    CAR = "car"
    TRUCK = "truck"
    PEDESTRIAN = "pedestrian"
    BICYCLE = "bicycle"
    SCOOTER = "scooter"


class RoutingMode(Enum):
    """Available RoutingMode Values."""

    SHORT = "short"
    FAST = "fast"


class UnitSystem(Enum):
    """Available UnitSystem Values."""

    METRIC = "metric"
    IMPERIAL = "imperial"


class Return(Enum):
    """Available Return Values."""

    POLYINE = "polyline"
    ACTIONS = "actions"
    INSTRUCTIONS = "instructions"
    SUMMARY = "summary"
    TRAVEL_SUMMARY = "travelSummary"
    ML_DURATION = "mlDuration"
    TYPICAL_DURATION = "typicalDuration"
    TURN_BY_TURN_ACTIONS = "turnByTurnActions"
    ELEVATION = "elevation"
    ROUTE_HANDLE = "routeHandle"
    PASSTHROUGH = "passthrough"
    INCIDENTS = "incidents"
    ROUTING_ZONES = "routingZones"
    TRUCK_ROAD_TYPES = "truckRoadTypes"
    TOLLS = "tolls"


class Spans(Enum):
    """Available Spans Values."""

    WALK_ATTRIBUTES = "walkAttributes"
    STREET_ATTRIBUTES = "streetAttributes"
    CAR_ATTRIBUTES = "carAttributes"
    TRUCK_ATTRIBUTES = "truckAttributes"
    SCOOTER_ATTRIBUTES = "scooterAttributes"
    NAMES = "names"
    LENGTH = "length"
    DURATION = "duration"
    BASE_DURATION = "baseDuration"
    TYPICAL_DURATION = "typicalDuration"
    COUNTRY_CODE = "countryCode"
    FUNCTIONAL_CLASS = "functionalClass"
    ROUTE_NUMBERS = "routeNumbers"
    SPEED_LIMIT = "speedLimit"
    MAX_SPEED = "maxSpeed"
    DYNAMIC_SPEED_INFO = "dynamicSpeedInfo"
    SEGMENT_ID = "segmentId"
    SEGMENT_REF = "segmentRef"
    CONSUMPTION = "consumption"
    ROUTING_ZONES = "routingZones"
    TRUCK_ROAD_TYPES = "truckRoadTypes"
    NOTICES = "notices"
    INCIDENTS = "incidents"
    TOLL_SYSTEMS = "tollSystems"


@dataclass
class Place:
    """Place for route requests."""

    latitude: float
    longitude: float


here_routing/here_routing.py
"""A module to query the HERE Routing API v8."""
from __future__ import annotations

import asyncio
import json
import socket
from datetime import datetime
from importlib import metadata
from typing import Any, List, MutableMapping

import aiohttp
import async_timeout
from yarl import URL

from .exceptions import (
    HERERoutingConnectionError,
    HERERoutingError,
    HERERoutingTooManyRequestsError,
    HERERoutingUnauthorizedError,
)
from .model import Place, Return, RoutingMode, Spans, TransportMode, UnitSystem

SCHEME = "https"
API_HOST = "router.hereapi.com"
API_VERSION = "/v8"
ROUTES_PATH = "routes"


class HERERoutingApi:
    """Main class for handling connections with the HERE Routing API v8."""

    def __init__(
        self,
        api_key: str,
        request_timeout: int = 10,
        session: aiohttp.client.ClientSession | None = None,
        user_agent: str | None = None,
    ) -> None:
        """Initialize connection with HERE Routing.
        Class constructor for setting up an HERERoutingApi object to
        communicate with the HERE Routing API.
        Args:
            api_key: A key generated specifically to authenticate API requests.
            request_timeout: Max timeout to wait for a response from the API.
            session: Optional, shared, aiohttp client session.
            user_agent: Defaults to here_routing/<version>.
        """
        self._session = session
        self._api_key = api_key
        self._close_session = False

        self.request_timeout = request_timeout
        self.user_agent = user_agent

        if user_agent is None:
            version = metadata.version(__package__)
            self.user_agent = f"here_routing/{version}"

    async def request(
        self,
        uri: str,
        params: MutableMapping[str, str | List[str]],
        method: str = "GET",
    ) -> Any:
        """Handle a request to the HERE Routing API.
        Make a request against the HERE Routing API and handles the response.
        Args:
            uri: The request URI on the HERE Routing API to call.
            method: HTTP method to use for the request; e.g., GET, POST.
            data: RAW HTTP request data to send with the request.
            json_data: Dictionary of data to send as JSON with the request.
            params: Mapping of request parameters to send with the request.
        Returns:
            The response from the API. In case the response is a JSON response,
            the method will return a decoded JSON response as a Python
            dictionary. In other cases, it will return the RAW text response.
        Raises:
            HERERoutingConnectionError: An error occurred while communicating
                with the HERE API (connection issues).
            HERERoutingError: An error occurred while processing the
                response from the HERE API (invalid data).
        """
        url = URL.build(scheme=SCHEME, host=API_HOST, path=API_VERSION) / uri

        headers = {
            "User-Agent": self.user_agent,
            "Accept": "application/json",
            "DNT": "1",
        }
        params["apiKey"] = self._api_key

        if self._session is None:
            self._session = aiohttp.ClientSession()
            self._close_session = True

        try:
            async with async_timeout.timeout(self.request_timeout):
                response = await self._session.request(
                    method,
                    url,
                    params=params,
                    headers=headers,
                )
        except asyncio.TimeoutError as exception:
            raise HERERoutingConnectionError(
                "Timeout occurred while connecting to the HERE Routing API."
            ) from exception
        except (aiohttp.ClientError, socket.gaierror) as exception:
            raise HERERoutingConnectionError(
                "Error occurred while communicating with the HERE Routing API."
            ) from exception

        content_type = response.headers.get("Content-Type", "")
        contents = await response.read()
        decoded_contents = contents.decode("utf8")
        if "application/json" not in content_type:
            raise HERERoutingError(response.status, {"message": decoded_contents})
        if response.status // 100 in [4, 5]:
            response.close()

            if response.status == 401:
                raise HERERoutingUnauthorizedError(
                    json.loads(decoded_contents)["error_description"]
                )
            if response.status == 429:
                raise HERERoutingTooManyRequestsError(
                    json.loads(decoded_contents)["error_description"]
                )

            raise HERERoutingError(response.status, json.loads(decoded_contents))

        return await response.json()

    async def route(  # pylint: disable=too-many-arguments
        self,
        transport_mode: TransportMode,
        origin: Place,
        destination: Place,
        routing_mode: RoutingMode = RoutingMode.FAST,
        alternatives: int = 0,
        units: UnitSystem = UnitSystem.METRIC,
        lang: str = "en-US",
        return_values: List[Return] | None = None,
        spans: List[Spans] | None = None,
        via: List[Place] | None = None,
        departure_time: datetime | None = None,
        arrival_time: datetime | None = None,
    ) -> Any:
        """Get the route.
        Args:
            transport_mode: The TransportMode to use.
            origin: Latitude and longitude of the origin.
            destination: Latitude and longitude of the destination.
            routing_mode: Routing Mode to use. Defaults to RoutingMode.FAST.
            alternatives: Number of alternative routes to return.
            units: Unitsystem to use.
            lang: IETF BCP 47 compatible language identifier.
            return_values: HERE Routing API return values to request.
            spans: Information for Spans to request.
            via: List of waypoints the route should route over.
            departure_time: Departure time.
            arrival_time: Arrival time.
        Returns:
            The response from the API.
        Raises:
            HERERoutingConnectionError: An error occurred while communicating
                with the HERE API (connection issues).
            HERERoutingError: An error occurred while processing the
                response from the HERE API (invalid data).
        """
        params: MutableMapping[str, str | List[str]] = {
            "transportMode": transport_mode.value,
            "origin": f"{origin.latitude},{origin.longitude}",
            "destination": f"{destination.latitude},{destination.longitude}",
            "routingMode": routing_mode.value,
            "alternatives": str(alternatives),
            "units": units.value,
            "lang": lang,
        }
        if return_values is not None:
            params["return"] = ",".join(r.value for r in return_values)
        if spans is not None:
            params["spans"] = ",".join(s.value for s in spans)
        if via is not None:
            params["via"] = [f"{v.latitude},{v.longitude}" for v in via]
        if departure_time is not None:
            params["departureTime"] = departure_time.isoformat(timespec="seconds")
        if arrival_time is not None:
            params["arrivalTime"] = arrival_time.isoformat(timespec="seconds")

        response = await self.request(uri=ROUTES_PATH, params=params)

        if len(response["routes"]) < 1:
            raise HERERoutingError(
                ",".join(notice["title"] for notice in response["notices"])
            )
        return response

    async def close(self) -> None:
        """Close open client session."""
        if self._session and self._close_session:
            await self._session.close()

    async def __aenter__(self) -> HERERoutingApi:
        """Async enter.
        Returns:
            The HERERoutingApi object.
        """
        return self

    async def __aexit__(self, *_exc_info) -> None:
        """Async exit.
        Args:
            _exc_info: Exec type.
        """
        await self.close()
